server/src/main/java/io/crate/types/NumericType.java

```diff
@@ -0,0 +1,237 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.types;
+
+import io.crate.Streamer;
+import io.crate.common.annotations.VisibleForTesting;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.math.MathContext;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class NumericType extends DataType<BigDecimal> implements Streamer<BigDecimal> {
+
+    public static final int ID = 22;
+    public static final NumericType INSTANCE = new NumericType(null, null); // unscaled
+
+    public static NumericType of(int precision) {
+        return new NumericType(precision, null);
+    }
+
+    public static NumericType of(int precision, int scale) {
+        return new NumericType(precision, scale);
+    }
+
+    public static DataType<?> of(List<Integer> parameters) {
+        if (parameters.isEmpty() || parameters.size() > 2) {
+            throw new IllegalArgumentException(
+                "The numeric type support one or two parameter arguments, received: " +
+                parameters.size()
+            );
+        }
+        if (parameters.size() == 1) {
+            return of(parameters.get(0));
+        } else {
+            return of(parameters.get(0), parameters.get(1));
+        }
+    }
+
+    private final Integer scale;
+    private final Integer precision;
+
+    private NumericType(Integer precision, Integer scale) {
+        this.precision = precision;
+        this.scale = scale;
+    }
+
+    public NumericType(StreamInput in) throws IOException {
+        this.precision = in.readOptionalVInt();
+        this.scale = in.readOptionalVInt();
+    }
+
+    @Override
+    public int id() {
+        return ID;
+    }
+
+    @Override
+    public Precedence precedence() {
+        return Precedence.NUMERIC;
+    }
+
+    @Override
+    public String getName() {
+        return "numeric";
+    }
+
+    @Override
+    public Streamer<BigDecimal> streamer() {
+        return this;
+    }
+
+    @Override
+    public BigDecimal implicitCast(Object value) throws IllegalArgumentException, ClassCastException {
+        if (value == null) {
+            return null;
+        }
+
+        var mathContext = mathContextOrDefault();
+        if (value instanceof BigDecimal) {
+            return new BigDecimal(
+                ((BigDecimal) value).toBigInteger(),
+                mathContext
+            ).setScale(scaleOrDefault(), mathContext.getRoundingMode());
```

Comment:
why not calling `.setScale` on the original value?

Reply:
We also have to set the precision, it is done via the math context.
If we set the scale on the original value, smth like that might result in an incorrect result 
`('some value'::numeric(16))::numeric(4, 2)`

Reply:
Ah right, I missed the precision part via the math context. Thanks for the explanation.